src/main/java/com/aws/greengrass/clientdevices/auth/iot/Thing.java

```diff
@@ -8,29 +8,143 @@
 import com.aws.greengrass.clientdevices.auth.session.attribute.AttributeProvider;
 import com.aws.greengrass.clientdevices.auth.session.attribute.DeviceAttribute;
 import com.aws.greengrass.clientdevices.auth.session.attribute.WildcardSuffixAttribute;
-import lombok.Value;
+import lombok.Getter;
 
+import java.util.ArrayList;
 import java.util.Collections;
+import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.regex.Pattern;
 
-@Value
-public class Thing implements AttributeProvider {
+/**
+ * This is a versioned representation of an IoT Thing. It is **NOT** updated
+ * when the local Thing Registry is updated, or when changes to this Thing are
+ * made in IoT Core. Consider calling the ThingRegistry to retrieve Thing
+ * objects as they are needed rather than storing references long term.
+ */
+@Getter
+public final class Thing implements AttributeProvider {
     public static final String NAMESPACE = "Thing";
     private static final String thingNamePattern = "[a-zA-Z0-9\\-_:]+";
 
-    String thingName;
+    private final int version;
+    private final String thingName;
+    private final List<String> attachedCertificateIds;
+    private boolean modified = false;
 
     /**
-     * Constructor.
+     * Create a new Thing.
+     *
      * @param thingName AWS IoT ThingName
      * @throws IllegalArgumentException If the given ThingName contains illegal characters
      */
-    public Thing(String thingName) {
+    public static Thing of(String thingName) {
+        return Thing.of(0, thingName);
+    }
+
+    /**
+     * Create a new Thing.
+     *
+     * @param version        Thing version
+     * @param thingName      AWS IoT ThingName
+     * @throws IllegalArgumentException If the given ThingName contains illegal characters
+     */
+    public static Thing of(int version, String thingName) {
+        return Thing.of(version, thingName, new ArrayList<>());
+    }
+
+    /**
+     * Create a new Thing.
+     *
+     * @param version        Thing version
+     * @param thingName      AWS IoT ThingName
+     * @param certificateIds Attached certificate IDs
+     * @throws IllegalArgumentException If the given ThingName contains illegal characters
+     */
+    public static Thing of(int version, String thingName, List<String> certificateIds) {
+        if (version < 0) {
```

Comment:
Thought: What do you thing about versions just being a wrapper (maybe we can do it later if we ever need to have more that 1 versions thing) but I was thinking about `Versioned<Thing>` and the api to interact with it would be something like

```java
Versioned<Thing> versionedThing = Versioned.of(thing)
Versioned<Thing> versionedThing = Versioned.of(10, thing)

int version = versionedThing.getVersion();
Thing thing = versionedThing.get();
```

Then the wrapper would have this validation logic. **again, this is thought** this works just fine.